### PR TITLE
Refine effects for getfield corner cases

### PR DESCRIPTION
**Pull request by Keno**
See individual commits.

### CODE REVIEW

**Review by vtjnash — APPROVED** [review]

File: test/compiler/effects.jl [hunk 1]
```diff
@@ -676,3 +676,11 @@ mksparamunused(x) = (SparamUnused(x); nothing)
...
     @test count(isnew, src.code) == 0
 end
+
+# Effects for getfield of type instance
+struct WrapperOneField{T}
+    x::T
+end
+Base.infer_effects(Tuple{Nothing}) do x
+    WrapperOneField{typeof(x)}.instance
+end |> Core.Compiler.is_total
```
Comment (ararslan) [inline test/compiler/effects.jl:686]:
```suggestion
@test Core.Compiler.is_total(Base.infer_effects(Tuple{Nothing}) do x
    WrapperOneField{typeof(x)}.instance
end)
```
I assume you meant to `@test` this?

Comment (Keno) [reply test/compiler/effects.jl:686]: Yes

**Comment by Keno** [comment]
I've pushed some tiny related commits to this PR, since otherwise, this'll just end up conflicting with each other. Should all be straightforward.

**Comment by aviatesk** [comment]
I will push some changes to fix test failures.

**Comment by Keno** [comment]
thanks

**Comment by Keno** [comment]
I'm concerned about that GC root issue in llvmpasses. I also saw something similar locally on a branch that included these changes. Is that currently a known issue on master or is that new on this branch? In either case, we should try to capture it in rr and see what the missing root is.

**Comment by aviatesk** [comment]
Maybe it's the issue that Jameson pointed out about #47994 (esp. https://github.com/JuliaLang/julia/pull/47994#issuecomment-1375664866)?

**Comment by Keno** [comment]
Assertions are enabled, so if that were the issue, i think we'd have seen the assertion in https://github.com/JuliaLang/julia/issues/48089.

**Comment by aviatesk** [comment]
(unrelated to the GC issue) So it seems like the remaining test failures (in `compiler/inline` and `compiler/inference`) are because this PR allows irinterp on more frames that are now proved as `:foldable` thanks to the `:consistent`-cy improvement of `getfield`.

**Comment by Keno** [comment]
The test in compiler/inline is fine, it's just getting some extra junk basic blocks:
```
CodeInfo(
1 ──       goto #3
2 ──       nothing::Nothing
3 ┄─ %3  = Base.getfield(x, 1)::Int64
└───       goto #4
4 ──       goto #5
5 ──       goto #7
6 ──       nothing::Nothing
7 ┄─ %8  = Base.getfield(x, 2)::Int64
└───       goto #8
8 ──       nothing::Nothing
└───       goto #10
9 ──       nothing::Nothing
10 ┄ %13 = Core.tuple(%3, %8)::Tuple{Int64, Int64}
└───       return %13
)
```
The test should be updated to test what it actually wants to know (two getfield calls, one tuple call, no other calls).

**Comment by aviatesk** [comment]
> The test in compiler/inline is fine, it's just getting some extra junk basic blocks:

Okay, just confirmed that LLVM optimizes those junks away.

The remaining ones are related to `MustAlias` lattice extension. I added it for JET's use case and it's not enabled in base yet. Since JET's analyzer doesn't enable irinterp, I'm fine with marking the failing cases as broken.

**Comment by aviatesk** [comment]
Okay, fixed up all the failed cases. Should be good to go.

**Pull request merged**